### ISSUE TITLE
Serialize topology object

### ISSUE THREAD

**Issue by andrrizzi** [pandegroup/openmm#1543]
As far as I understand, the `Topology` object cannot be serialized through `XMLSerializer` like `System`s. Would it be possible/make sense to add this?

Currently, the only "safe" way I could find to serialize was to go through `mdtraj`, transform the `Topology` into a `pandas` `DataFrame` object, and serialize that as JSON string. This way I'll be able to read old files as long as the API of `mdtraj.Topology.to_dataframe()` doesn't change, and I don't have to rely on `pickle` which comes with poor future compatibility once the internal implementation changes. It would be nice to have a more robust way to do this though.

**Comment by proteneer** [comment]
The difference is that the System class is a C++ object, and the Topology is a python-specific class. The core XML serializers are all written in C++ for C++ objects. No such mechanism currently exists for Python objects. But maybe @peastman has better ideas :)

**Comment by mpharrigan** [comment]
You're right that `pickle` has poor future (forwards) compatibility. But it does have backwards compatibility, which is probably more important here. The use case for forwards compatibility is serializing something from python3 and loading it in python2. This is still a solvable problem: on py3 use
```python
pickle.dump(obj, f, protocol=2)
```

**Comment by andrrizzi** [comment]
Thank you all for your answers!

Sorry if I wasn't clear. I'm actually more worried about future changes in `Topology` internal implementation. To make a minimal example

```python
class TopologyMini(object):

    def __init__(self):
        self._chains = []

    def getNumChains(self):
        return len(self._chains)

top_mini = TopologyMini()
top_mini.getNumChains()
with open('top.pickle', 'w') as f:
    pickle.dump(top_mini, f)
```
if in the future we change the internal representation of `TopologyMini` to
```python
class TopologyMini(object):

    def __init__(self):
        self._chains = []
        self._numChains = 0

    def getNumChains(self):
        return self._numChains

with open('top.pickle', 'r') as f:
    top_mini = pickle.load(f)
top_mini.getNumChains()  # Fail even if the API is the same!
```
the last line raises `AttributeError: 'TopologyMini' object has no attribute '_numChains'`.

This is why I think it would be nice to have a representation of the `Topology` object state that is independent of its implementation. For simple `Python` objects I believe this is as simple as overriding the `__getstate__` and `__setstate__` methods.

**Comment by mpharrigan** [comment]
Sorry! Yes, that is a good point

**Comment by jchodera** [comment]
I just wanted to again suggest that making the pure Python objects in OpenMM's `app` layer serializable would make everybody's lives a lot easier. I just spent a few hours wrestling with writing JSON serializer/deserializers for components of `ForceField`, but there's no sense in everyone having to do this independently when then need to serialize our Python objects. 

Perhaps we could identify which classes we can target for serialization?

**Comment by peastman** [comment]
Which ones do you suggest?  Topology and ForceField are the most obvious ones.  In principle we could do reporters, but I'm not sure how much value there would be in that.

**Comment by peastman** [comment]
This came up again in #3202.  Let's discuss what would be involved in providing better serialization for Python objects.  First we need to decide what classes should be serializable.  Here are the main ones I can think of to consider.

- Vec3
- Unit
- Quantity
- Topology
- ForceField
- Python subclasses of C++ classes (i.e. CustomIntegrator subclasses)

The first four should be easy except for one major issue: to serialize a Quantity, you also need to serialize whatever value it wraps.  That's easy for primitive types, but may be harder for numpy arrays.

ForceField is a lot more difficult.  It's large, complicated, modular, and extensible.

Python subclasses of C++ classes could be very difficult.  But we could at least investigate it to see what's possible.

Next we need to consider what serialization format(s) to support.  Here are some of the most obvious ones.

- Pickle.  This isn't good for long term persistence, but it's important for some other purposes.  For example, `multiprocessing` uses pickling to transfer objects between processes.
- XML.  This is what all the existing serialization is based on.  For consistency, it should be a top candidate.
- JSON.  Also a very common serialization format.

Once you have one serialization format, that can often be used to easily implement others.  For example, we implement pickling for C++ objects by just serializing them to XML and returning that from `__getstate__()`.  In principle you could also go the other way, using `pickle.dumps()` to produce a string that you embed in a XML or JSON file.  That wouldn't be ideal, though, since it would have pickling's lack of forward compatibility.

**Comment by mattwthompson** [comment]
I don't like bumping old issues but this is a pretty important one and I have some experience that might be worth sharing.

As somebody who might be rolling their own function that dumps a `Topology` to a JSON representation, being able to get these objects into dicts of only primitive types does 99% of the work of JSON serialization. I assume it also gets most of the way to the pickle and XML solutions. It also picks up some other serialization formats that seem good but don't appear to have much adoption in computational chemistry (BSON, MessagePack, probably a bunch of others that web devs like to use).

> The first four should be easy except for one major issue: to serialize a Quantity, you also need to serialize whatever value it wraps. That's easy for primitive types, but may be harder for numpy arrays.

We have run into this precisely as you've described and have been experimenting with a solution that has worked so far. It dumps the `Quantity` object to a dictionary that separately tracks the unit (as a string) and whatever numerical values the `Quantity` object wraps. You can see it in action [here](https://github.com/mattwthompson/openff-models/blob/61260ec1d2746a0431671498dfe1540b006354db/openff/models/types.py#L105-L119). This is with a Pint registry but probably works similarly enough for `Vec3` and such, maybe with some added [logic to recurse](https://github.com/openforcefield/openff-toolkit/blob/5bd279c7a0cb7b309b5c2087828cfbabc5744cea/openff/toolkit/utils/serialization.py#L485) through more nested dicts.

There's obviously some room to improve compression, etc. at the cost of human-readability. We just haven't scaled up this solution to stuff like millions of atoms, so there hasn't been much of a need to care about that sort of performance.
